js/idex.js

```diff
@@ -0,0 +1,969 @@
+            const signature = this.signMessage (orderHash, this.privateKey);
+            const request = {
+                'tokenBuy': tokenBuy,
+                'amountBuy': amountBuy,
+                'tokenSell': tokenSell,
+                'amountSell': amountSell,
+                'address': this.walletAddress,
+                'nonce': nonce,
+                'expires': expires,
+            };
+            const response = await this.privatePostOrder (this.extend (request, signature)); // this.extend (request, params) will cause invalid signature
+            // { orderNumber: 1562323021,
+            //   orderHash:
+            //    '0x31c42154a8421425a18d076df400d9ec1ef64d5251285384a71ba3c0ab31beb4',
+            //   timestamp: 1564041428,
+            //   price: '0.00073',
+            //   amount: '210',
+            //   total: '0.1533',
+            //   type: 'buy',
+            //   params:
+            //    { tokenBuy: '0x763fa6806e1acf68130d2d0f0df754c93cc546b2',
+            //      buyPrecision: 18,
+            //      amountBuy: '210000000000000000000',
+            //      tokenSell: '0x0000000000000000000000000000000000000000',
+            //      sellPrecision: 18,
+            //      amountSell: '153300000000000000',
+            //      expires: 100000,
+            //      nonce: 1,
+            //      user: '0x0ab991497116f7f5532a4c2f4f7b1784488628e1' } }
+            return this.parseOrder (response, market);
+        } else if (type === 'market') {
+            if (!('orders' in params)) {
+                throw new ArgumentsRequired (this.id + ' market order requires a list orders as returned from idex.privatePostReturnOrderBook');
+            }
+            const orders = params['orders'];
+            const request = [];
+            for (let i = 0; i < orders.length; i++) {
+                const order = orders[i];
+                // { price: '0.000132247803328924',
+                //   amount: '19980',
+                //   total: '2.6423111105119',
+                //   orderHash:
+                //    '0x5fb3452b3d13fc013585b51c91c43a0fbe4298c211243763c49437848c274749',
+                //   params:
+                //    { tokenBuy: '0x0000000000000000000000000000000000000000',
+                //      buySymbol: 'ETH',
+                //      buyPrecision: 18,
+                //      amountBuy: '2642311110511900000',
+                //      tokenSell: '0xb705268213d593b8fd88d3fdeff93aff5cbdcfae',
+                //      sellSymbol: 'IDEX',
+                //      sellPrecision: 18,
+                //      amountSell: '19980000000000000000000',
+                //      expires: 10000,
+                //      nonce: 1564656561510,
+                //      user: '0xc3f8304270e49b8e8197bfcfd8567b83d9e4479b' } }
+                const orderToSign = {
+                    'orderHash': order['orderHash'],
+                    'amount': order['params']['amountBuy'],
```

Comment:
Tbh, I don't think that users will really bother to specify exchange-specific arguments to a unified method, which would then unwrap those exchange-specific values and send them to the endpoint. If the lib requires going down to a level that low (down to calculating specific values in wei/decimals) – then it is easier to just use the implicit endpoint directly. So, in this case, we have two engineering options:

1. Accept some kind of an array of unified order structures (`[{ id, amount }, ...]`) and do the precision stuff for the user (then the method does useful unification work for the user and the user has reasons to call the method). Calculate everything automatically as much as possible (`nonce`, and, maybe, some other fields).

2. This option i like more: do not attempt to cross the orderbook, handle one order at a time (no loops required). Having decided that, the method for placing market orders then should not expect `params['orders']`, instead it would require `params['orderHash']` and it will probably simplify a lot, because the code would be very close if not identical to the `limit` part.

Reply:
I think that it would be easiest for users to use the exchange specific arguments, because otherwise they would have to parse the data from the orderbook. Accepting `params['orderHash']` would not work because the endpoint requires more data to be signed, including the amount and the hash, so we would have to make another request to get the orderbook and find a matching orderhash.

Reply:
> Accepting `params['orderHash']` would not work because the endpoint requires more data to be signed, including the amount and the hash, so we would have to make another request to get the orderbook and find a matching orderhash.

Uhm, nope, suppose the user already did this:

> make another request to get the orderbook and find a matching orderhash

(A user will only want to place an order after fetching the orderbook, so we don't have to re-fetch it again inside `createOrder`).

In that case, the user supplies the `orderHash` in params, and the `amount` is taken from the args, and the `nonce` is fetched just like we do now with `limit` orders, so, it boils down to two linear calls that we can't break down into smaller pieces: 1. fetch the nonce, 2. submit the order. Lmk if you agree.